truncate/table_schema.go

```diff
@@ -137,6 +118,93 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
 	return tables, nil
 }
 
+// filterTableSchemas filters tables with given targetTables and excludeTables.
+// If targetTables is not empty, it fetches only the specified tables.
+// If excludeTables is not empty, it excludes the specified tables.
+// TargetTables and excludeTables cannot be specified at the same time.
+func filterTableSchemas(tables []*tableSchema, targetTables, excludeTables []string) ([]*tableSchema, error) {
+	isExclude := len(excludeTables) > 0
+	isTarget := len(targetTables) > 0
+
+	switch {
+	case isTarget && isExclude:
+		return nil, errors.New("both targetTables and excludeTables cannot be specified at the same time")
+	case isTarget && !isExclude:
+		return targetFilterTableSchemas(tables, targetTables), nil
+	case !isTarget && isExclude:
+		return excludeFilterTableSchemas(tables, excludeTables), nil
+	default: // No target and exclude tables are specified.
+		return tables, nil
+	}
+}
+
+// targetFilterTableSchemas filters tables with given targetTables.
+// If targetTables is empty, it returns all tables.
+func targetFilterTableSchemas(tables []*tableSchema, targetTables []string) []*tableSchema {
+	if len(targetTables) == 0 {
+		return tables
+	}
+
+	targets := make(map[string]struct{}, len(targetTables))
+	for _, t := range targetTables {
+		targets[t] = struct{}{}
+	}
+
+	filtered := make([]*tableSchema, 0, len(tables))
+	for _, t := range tables {
+		if _, ok := targets[t.tableName]; ok {
+			filtered = append(filtered, t)
+		}
+	}
+
+	return filtered
+}
+
+// excludeFilterTableSchemas filters tables with given excludeTables.
+// If excludeTables is empty, it returns all tables.
+// When an exclude table is cascade deletable, its parent table is also excluded.
+func excludeFilterTableSchemas(tables []*tableSchema, excludeTables []string) []*tableSchema {
+	if len(excludeTables) == 0 {
+		return tables
+	}
+
+	excludes := make(map[string]struct{}, len(tables))
+	for _, t := range excludeTables {
+		excludes[t] = struct{}{}
+	}
+
+	// Add parent tables that may delete the exclude tables in cascade
+	// Since interleave tables can be hierarchical, tracing up to the top level is needed.
+	for {
+		excludeParents := make(map[string]struct{}, len(excludes))
+		for _, t := range tables {
+			if _, ok := excludes[t.tableName]; ok && t.isCascadeDeletable() {
+				if _, alreadyExcluded := excludes[t.parentTableName]; !alreadyExcluded {
+					excludeParents[t.parentTableName] = struct{}{}
+				}
+			}
+		}
+
+		// loop until no more parent tables to exclude
+		if len(excludeParents) == 0 {
+			break
+		}
+
+		for p := range excludeParents {
+			excludes[p] = struct{}{}
+		}
+	}
```

Comment:
hmm... this would a bit hard to comprehend how it works.

How about splitting the entire logic into two steps - 1) find the all excluded tables recursively, and then 2) exclude tables?

In pseudo code:
```go
func findExcludeTables(tables []*tableSchema, excludeTables []string) []string {
	excludes := make(map[string]bool, len(tables))
	for _, t := range excludeTables {
		excludes[t] = true
	}

	var allExcludes []string
	for _, t := range tables {
		if _, ok := excludes[t.tableName]; ok {
			allExcludes = append(allExcludes, t.tableName)
			if t.isCascadeDeletable() {
				excluded := findExcludeTables(tables, []string{t.parentTableName})
				allExcludes = append(allExcludes, excluded...)
			}
		}
	}
	return allExcludes
}

func excludeFilterTableSchemas(tables []*tableSchema, excludeTables []string) []*tableSchema {
	excludes := findExcludeTables(tables, excludeTables)
	...
}
```